engine/consensus/sealing/core.go

```diff
@@ -112,6 +113,84 @@ func NewCore(
 	return core, nil
 }
 
+// RepopulateAssignmentCollectorTree restores latest state of assignment collector tree based on local chain state information.
+// Repopulating is split into two parts:
+// 1) traverse forward all finalized blocks starting from last sealed block till we reach last finalized block . (lastSealedHeight, lastFinalizedHeight]
+// 2) traverse forward all unfinalized(pending) blocks starting from last finalized block.
+// For each block that is being traversed we will collect execution results and process them using sealing.Core.
+func (c *Core) RepopulateAssignmentCollectorTree(payloads storage.Payloads) error {
+	finalizedSnapshot := c.state.Final()
+	finalized, err := finalizedSnapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not retrieve finalized block: %w", err)
+	}
+	finalizedID := finalized.ID()
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork.
+	latestSeal, err := c.seals.ByBlockID(finalizedID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", finalizedID, err)
+	}
+
+	latestSealedBlockID := latestSeal.BlockID
+	latestSealedBlock, err := c.headers.ByBlockID(latestSealedBlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve latest sealed block (%x): %w", latestSealedBlockID, err)
+	}
+
+	// usually we start with empty collectors tree, prune it to minimum height
+	_, err = c.collectorTree.PruneUpToHeight(latestSealedBlock.Height)
+	if err != nil {
+		return fmt.Errorf("could not prune execution tree to height %d: %w", latestSealedBlock.Height, err)
+	}
+
+	// resultProcessor adds _all known_ results for the given block to the assignment collector tree
+	resultProcessor := func(header *flow.Header) error {
+		payload, err := payloads.ByBlockID(header.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve index for block (%x): %w", header.ID(), err)
+		}
+
+		for _, result := range payload.Results {
+			// TODO: change this when migrating to sealing & verification phase 3.
+			// Incorporated result is created this way only for phase 2.
+			incorporatedResult := flow.NewIncorporatedResult(result.BlockID, result)
+			err = c.ProcessIncorporatedResult(incorporatedResult)
+			if err != nil {
+				return fmt.Errorf("could not process incorporated result: %w", err)
+			}
+		}
+		return nil
+	}
+
+	// traverse chain forward to collect all execution results that were incorporated in this fork
+	// starting from finalized block and finishing with latest sealed block
+	err = fork.TraverseForward(c.headers, finalizedID, resultProcessor, fork.ExcludingBlock(latestSealedBlockID))
+	if err != nil {
+		return fmt.Errorf("internal error while traversing fork: %w", err)
+	}
+
+	// at this point we have processed all results in range (lastSealedBlock, lastFinalizedBlock].
+	// Now, we add all known results for any valid block that descends from the latest finalized block:
```

Comment:
Nice to have these comments. I would suggest to add some info log for the progress. 

As the repopulation might take some time, it's useful to log the how many blocks we are need traverse in total, and the progress. 

For instance, show logs like:
```
"there are 100 finalized and unsealed blocks in total to reload into collector tree with assignment"
"20/100 have been loaded to collector tree"
"40/100 have been loaded to collector tree"
...
"100/100 have been loaded to collector tree"
"there are 50 unfinalized blocks to load into the collector tree with assignment"
"20/50 have been loaded to collector tree"
"40/50 have been loaded to collector tree"
"50/50 have been loaded to collector tree"
```

A simple way is to print a log for every 20 blocks, if these steps takes some time, the log would give us a better idea how long it takes to finish the reload.